src/core/vectortile/qgsmapboxglstyleconverter.cpp

```diff
@@ -1938,24 +1992,130 @@ QImage QgsMapBoxGlStyleConverter::retrieveSprite( const QString &name, QgsMapBox
   return sprite;
 }
 
-QString QgsMapBoxGlStyleConverter::retrieveSpriteAsBase64( const QString &name, QgsMapBoxGlStyleConversionContext &context, QSize &size )
+QString QgsMapBoxGlStyleConverter::retrieveSpriteAsBase64( const QVariant &value, QgsMapBoxGlStyleConversionContext &context, QSize &size, QString &spriteProperty, QString &spriteSizeProperty )
 {
-  const QImage sprite = retrieveSprite( name, context );
-  if ( !sprite.isNull() )
+  QString returnPath;
+  switch ( value.type() )
   {
-    size = sprite.size();
-    QByteArray blob;
-    QBuffer buffer( &blob );
-    buffer.open( QIODevice::WriteOnly );
-    sprite.save( &buffer, "PNG" );
-    buffer.close();
-    QByteArray encoded = blob.toBase64();
+    case QVariant::String:
+    {
+      spriteProperty.clear();
+      spriteSizeProperty.clear();
+      const QImage sprite = retrieveSprite( value.toString(), context );
+      if ( !sprite.isNull() )
+      {
+        size = sprite.size() * context.pixelSizeConversionFactor();
+        QByteArray blob;
+        QBuffer buffer( &blob );
+        buffer.open( QIODevice::WriteOnly );
+        sprite.save( &buffer, "PNG" );
+        buffer.close();
+        QByteArray encoded = blob.toBase64();
+
+        QString path( encoded );
+        path.prepend( QLatin1String( "base64:" ) );
+        returnPath = path;
+      }
+      break;
+    }
+
+    case QVariant::Map:
+    {
+      const QVariantList stops = value.toMap().value( QStringLiteral( "stops" ) ).toList();
+      if ( stops.size() == 0 )
+        break;
+
+      QString path;
+      QImage sprite;
 
-    QString path( encoded );
-    path.prepend( QLatin1String( "base64:" ) );
-    return path;
+      sprite = retrieveSprite( stops.value( 0 ).toList().value( 1 ).toString(), context );
+      if ( !sprite.isNull() )
+      {
+        size = sprite.size() * context.pixelSizeConversionFactor();
+        QByteArray blob;
+        QBuffer buffer( &blob );
+        buffer.open( QIODevice::WriteOnly );
+        sprite.save( &buffer, "PNG" );
+        buffer.close();
+        QByteArray encoded = blob.toBase64();
+
+        path = QString( encoded );
+        path.prepend( QLatin1String( "base64:" ) );
+        returnPath = path;
+      }
+      spriteProperty = QStringLiteral( "CASE WHEN @zoom_level < %1 THEN '%2'" )
+                       .arg( stops.value( 0 ).toList().value( 0 ).toString() )
+                       .arg( path );
+      spriteSizeProperty = QStringLiteral( "CASE WHEN @zoom_level < %1 THEN %2" )
+                           .arg( stops.value( 0 ).toList().value( 0 ).toString() )
+                           .arg( size.width() );
+
+      QSize s;
+      for ( int i = 0; i < stops.size() - 1; ++i )
+      {
+        ;
+        sprite = retrieveSprite( stops.value( 0 ).toList().value( 1 ).toString(), context );
+        if ( !sprite.isNull() )
+        {
+          s = sprite.size() * context.pixelSizeConversionFactor();
```

Comment:
This same code is used multiple times - could you pull it out to a single function?

Reply:
Will do.